RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -16,37 +20,64 @@
 import reactor.core.publisher.Flux;
 
 @RequiredArgsConstructor
+@Slf4j
 public class LuceneDocumentsReader {
-    private static final Logger logger = LogManager.getLogger(LuceneDocumentsReader.class);
-    protected final Path luceneFilesBasePath;
+    public static final int NUM_DOCUMENTS_BUFFERED = 1024;
+    protected final Path indexDirectoryPath;
 
-    public Flux<Document> readDocuments(String indexName, int shardId) {
-        Path indexDirectoryPath = luceneFilesBasePath.resolve(indexName).resolve(String.valueOf(shardId));
-
-        return Flux.using(
-            () -> openIndexReader(indexDirectoryPath),
-            reader -> {
-                logger.info(reader.maxDoc() + " documents found in the current Lucene index");
+    public class DocumentOrDone {}
+    @Getter
+    @AllArgsConstructor
+    public class DocumentHolder extends DocumentOrDone {
+        Document value;
+    }
+    public class Done extends DocumentOrDone {}
 
-                return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
-                .handle((i, sink) -> {
-                    Document doc = getDocument(reader, i);
+    public BlockingQueue<DocumentOrDone> packageDocumentsIntoQueue() {
+        var docQueue = new ArrayBlockingQueue<DocumentOrDone>(NUM_DOCUMENTS_BUFFERED, true);
+        new Thread(() -> {
+            try (var reader = openIndexReader(indexDirectoryPath)) {
+                log.info(reader.maxDoc() + " documents found in the current Lucene index");
+                for (var i=0; i<reader.maxDoc(); ++i) {
+                    var doc = getDocument(reader, i);
                     if (doc != null) { // Skip malformed docs
-                        sink.next(doc);
+                        docQueue.add(new DocumentHolder(doc));
                     }
-                }).cast(Document.class);
-            },
-            reader -> { // Close the IndexReader when done
+                }
+                docQueue.add(new Done());
+            } catch (IOException e) {
+                throw Lombok.sneakyThrow(e);
+            }
+        }).start();
+        return docQueue;
+    }
+
+    public Flux<Document> readDocuments() {
+        var queue = packageDocumentsIntoQueue();
+        return Flux.create(sink -> {
+            while (true) {
                 try {
-                    reader.close();
-                } catch (IOException e) {
-                    logger.error("Failed to close IndexReader", e);
-                    throw Lombok.sneakyThrow(e);
+                    var item = queue.take(); // TODO - this is still a blocking call!
+                    if (item instanceof Done) {
+                        sink.complete();
+                        break;
+                    }
+                    sink.next(((DocumentHolder)item).value);
+                } catch (InterruptedException e) {
```

Comment:
Lets also log this when it occurs

Reply:
See the above comment - not relevant any more.